src/python/pants/engine/target.py

```diff
@@ -22,25 +24,55 @@ def __repr__(self) -> str:
 class PrimitiveField(Field, metaclass=ABCMeta):
     """A Field that does not need the engine in order to be hydrated.
 
-    This should be subclassed by the majority of fields.
+    This should be used by the majority of fields.
+
+    Subclasses must implement `hydrate()` to convert `self.raw_value` into `self.value`. This
+    hydration and/or validation happens eagerly in the constructor. If the hydration is
+    particularly expensive, use `AsyncField` instead to get the benefits of the engine's caching.
+
+    Subclasses should also override the type hints for `raw_value` and `value` to be more precise
+    than `Any`.
+
+    Example:
+
+        class ZipSafe(PrimitiveField):
+            alias: ClassVar = "zip_safe"
+            raw_value: Optional[bool]
+            value: bool
+
+            def hydrate(self) -> bool:
+                if self.raw_value is None:
+                    return True
+                return self.raw_value
     """
 
-    def __str__(self) -> str:
-        return f"{self.alias}={self.value}"
+    value: Any
+
+    @final
+    def __init__(self, raw_value: Optional[Any]) -> None:
+        self.raw_value = raw_value
+        self.value = self.hydrate()
 
-    @memoized_property
     @abstractmethod
-    def value(self) -> Any:
-        """Get the field's value.
+    def hydrate(self) -> Any:
+        """Convert `self.raw_value` into `self.value`.
 
-        The value will possibly be first hydrated and/or validated, such as using a default value
-        if the field was not defined or ensuring that an int value is positive.
+        You should perform any validation and/or hydration here. For example, you may want to check
+        that an integer is > 0, apply a default value if `raw_value` is None, or convert an
+        Iterable[str] to List[str].
 
-        This property is memoized because hydration and validation can often be costly. This
-        hydration is lazy, i.e. it will only happen when a downstream rule explicitly requests this
-        field.
+        If you have no validation/hydration, simply set this function to `return self.raw_value`.
```

Comment:
Alternatively, we could make it so this is not an abstract method. But I very intentionally set this to be abstract so that every field author must consciously decide "do I want to do any validation or hydration?"

The vast majority of `Field`s will never end up implementing this method because they will subclass fields like `BoolField`, `StringField`, `IntField`, etc, that do all the hydration for them.